### ISSUE TITLE
Question: train.py

### ISSUE THREAD

**Issue by mohamedelbahnasawi** [hkproj/pytorch-transformer#5]
Hi @hkproj, I have a question regarding the train loop, In dataset.py you added a <SOS> token to the decoder input which means that the model prediction will include this token at the beginning of the prediction. However, in the label tensor, there is no <SOS> token. My question is, when calculating the loss in lines [223:229] in train.py :
```
proj_output = model.project(decoder_output) # (B, seq_len, vocab_size)

# Compare the output with the label
  label = batch['label'].to(device) # (B, seq_len)

  # Compute the loss using a simple cross-entropy
 loss = loss_fn(proj_output.view(-1, tokenizer_tgt.get_vocab_size()), label.view(-1))
 
```
 Wouldn't be a misalignment in the loss calculation and we need to calculate the loss for predictions and labels?
 
should we slice off the predictions by 1 and the labels by -1 ?

```
proj_output = proj_output[:, 1:, :]
label = batch['label'].to(device)[:, :-1]
```

**Comment by hkproj** [comment]
Hi @mohamedelbahnasawi 

"In dataset.py you added a token to the decoder input which means that the model prediction will include this token at the beginning of the prediction".

This is not how it works. The decoder's input is shifted compared to the target. Why? Because the transformer is a sequence-to-sequence model, so it will learn how to map the input sequence to the output sequence, but of course we want the `SOS` token to map to the first translated word, not to the `SOS` token itself (it would make no sense, since we wouldn't know how to get the next token then).

For example in the following English->French translation task, considering a sentence length of 10 and each word being a token, we have:

Encoder Input: `[SOS] My sentence in English [EOS] [PAD] [PAD] [PAD] [PAD]`
Decoder Input: `[SOS] ma phrase en français [PAD] [PAD] [PAD] [PAD] [PAD]` 
Target: `ma phrase en français [EOS] [PAD] [PAD] [PAD] [PAD] [PAD]`

The transformer will map the `[SOS]` token as the input of the decoder to the first token of the french translation `ma`. When given `[SOS] ma` as input to the decoder, it will be mapped to the output sequence `ma phrase`. So, given `N` tokens as input to the decoder input, will return `N` tokens as output of the decoder.

When we evaluate the loss, we calculate the cross attention between the target and the models' output. The only thing we ignore when calculating the loss is the [PAD] token, for which we specify the ignore_index when allocating the CrossEntropyLoss class: `loss_fn = nn.CrossEntropyLoss(ignore_index=tokenizer_src.token_to_id('[PAD]'), label_smoothing=0.1).to(device)`

I suggest you watch my video on the models' explanation, especially the part about training and inference: https://youtu.be/bCz4OMemCcA?t=2701

Have a nice day

**Comment by mohamedelbahnasawi** [comment]
Great! Thank you for the clarification. I watched the video and understood what you mean. I think I got the confusion from the `greedy_decode `function because `model_out` always include the <SOS> token in the beginning. So, I think to calculate the validation accuracy for example based on the label and model_out tensors, in this case would we need to shift the `model_out `to the right?

**Comment by hkproj** [comment]
Hi @mohamedelbahnasawi 
The fact that `model_output` has a [SOS] token at the beginning is because in the function `greedy_decode`, I have been using the decoder input also as the output variable, since I always append the last output token to it. To avoid confusion, I should have used another variable (basically a copy of the `decoder_input` variable in the `greedy_decode` function, but without the `[SOS` token).

Yes, in order to evaluate the model's output against the label, you should remove that `[SOS]` token at the beginning.

**Comment by mohamedelbahnasawi** [comment]
Everything makes more sense for me now! Thank you very much and have a nice day!
